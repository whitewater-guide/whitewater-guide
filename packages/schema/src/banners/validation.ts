import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { array, bool, mixed, number, object, string } from 'yup';

import type { BannerInput, BannerSourceInput } from '../__generated__/types';
import { BannerKind, BannerPlacement } from '../__generated__/types';

const BannerSourceSchema: ObjectSchema<BannerSourceInput> = object({
  kind: mixed<BannerKind>().oneOf(Object.values(BannerKind)).required(),
  url: string().required(),
}).noUnknown();

export const BannerInputSchema: ObjectSchema<BannerInput> = object({
  id: string().uuid().nullable(),
  slug: string().slug().required(),
  name: string().required(),
  priority: number().integer().required(),
  enabled: bool().required(),
  placement: mixed<BannerPlacement>()
    .oneOf(Object.values(BannerPlacement))
    .required(),
  source: BannerSourceSchema.clone(),
  link: string().url().https().nullable(),
  extras: object().nullable(),
  regions: array(yupSchemas.refInput().required()).required(),
  groups: array(yupSchemas.refInput().required()).required(),
})
  .strict(true)
  .noUnknown(true);
