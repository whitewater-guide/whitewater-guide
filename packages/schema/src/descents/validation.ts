import type { ObjectSchema } from 'yup';
import { bool, date, number, object, string } from 'yup';

import type { DescentInput, DescentLevelInput } from '../__generated__/types';

export const DescentLevelInputSchema: ObjectSchema<DescentLevelInput> = object({
  unit: string().max(20).optional().nullable(),
  value: number().required(),
})
  .strict(true)
  .noUnknown();

export const DescentInputSchema: ObjectSchema<DescentInput> = object({
  id: string().uuid().optional().nullable(),
  sectionId: string().uuid().required().nonNullable(),
  startedAt: date().required(),
  duration: number().integer().min(0).notRequired().nullable(),
  level: DescentLevelInputSchema.clone().notRequired().nullable(),
  comment: string().notRequired().nullable(),
  public: bool().notRequired().nullable(),
})
  .strict(true)
  .noUnknown();
