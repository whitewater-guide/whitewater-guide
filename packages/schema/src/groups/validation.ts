import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { GroupInput } from '../__generated__/types';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema: ObjectSchema<GroupInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nonEmpty(),
  sku: string()
    .defined()
    .max(255)
    .matches(GROUP_SKU, 'yup:string.sku')
    .nullable(),
})
  .strict(true)
  .noUnknown();
