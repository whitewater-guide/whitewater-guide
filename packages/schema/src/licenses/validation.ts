import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { LicenseInput } from '../__generated__/types';

export const LicenseInputSchema: ObjectSchema<LicenseInput> = object({
  url: string().url().nullable(),
  name: string().min(2).defined().nonNullable(),
  slug: string().slug().nullable(),
})
  .strict(true)
  .noUnknown();
