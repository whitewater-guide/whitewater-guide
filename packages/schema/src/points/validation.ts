import { isArray, isNumber } from 'lodash';
import type { ObjectSchema, Schema } from 'yup';
import { array, object, string } from 'yup';

import type { PointInput } from '../__generated__/types';
import { POITypes } from './POITypes';

export const CoordinateSchema: Schema<CodegenCoordinates> = array()
  .max(3)
  .test({
    name: 'is-coordinate3d-lng',
    test(v) {
      const valid = isArray(v) && isNumber(v[0]) && v[0] >= -180 && v[0] <= 180;
      const path = this.path ? `${this.path}.0` : '0';
      return (
        valid || this.createError({ path, message: 'yup:number.longitude' })
      );
    },
  })
  .test({
    name: 'is-coordinate3d-lat',
    test(v) {
      const valid = isArray(v) && isNumber(v[1]) && v[1] >= -90 && v[1] <= 90;
      const path = this.path ? `${this.path}.1` : '1';
      return (
        valid || this.createError({ path, message: 'yup:number.latitude' })
      );
    },
  })
  .test({
    name: 'is-coordinate3d-alt',
    test(v) {
      const path = this.path ? `${this.path}.2` : '2';
      return (
        (isArray(v) &&
          (v[2] === undefined || v[2] === null || isNumber(v[2]))) ||
        this.createError({ path, message: 'yup:number.altitude' })
      );
    },
  }) as any;

export const PointInputSchema: ObjectSchema<PointInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nullable(),
  description: string().nullable(),
  coordinates: CoordinateSchema.clone(),
  kind: string().defined().oneOf(POITypes),
})
  .strict(true)
  .noUnknown();
