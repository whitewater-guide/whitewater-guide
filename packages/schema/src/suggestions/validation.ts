import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { array, number, object, string } from 'yup';

import type { SuggestionInput } from '../__generated__/types';

function makeSuggestingSchema(
  photoRequired: boolean,
): ObjectSchema<SuggestionInput> {
  const description = photoRequired
    ? string().notRequired()
    : string().required();
  const filename = photoRequired ? string().required() : string().notRequired();
  const resolutionArr = array()
    .of(number().integer().positive().required())
    .min(2)
    .max(2);
  const resolution = photoRequired
    ? resolutionArr.required()
    : resolutionArr.notRequired();

  return object({
    section: yupSchemas.refInput().defined(),
    description,
    copyright: string().nullable(),
    filename,
    resolution,
  })
    .test({
      name: 'filename_and_resolution',
      test(v) {
        if (!!v.filename && !v.resolution) {
          return this.createError({
            path: 'resolution',
            message: 'yup:mixed.required',
          });
        }
        if (!v.filename && !!v.resolution) {
          return this.createError({
            path: 'filename',
            message: 'yup:mixed.required',
          });
        }
        return true;
      },
    })
    .strict(true)
    .noUnknown();
}

export const SuggestionInputSchema: ObjectSchema<SuggestionInput> =
  makeSuggestingSchema(false);

export const PhotoSuggestionInputSchema: ObjectSchema<SuggestionInput> =
  makeSuggestingSchema(true);
