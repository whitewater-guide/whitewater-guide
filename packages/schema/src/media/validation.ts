import type { ObjectSchema } from 'yup';
import { mixed, number, object, string, tuple } from 'yup';

import type { MediaInput } from '../__generated__/types';
import { MediaKind } from '../__generated__/types';
import { LicenseInputSchema } from '../licenses';

export const MediaInputSchema: ObjectSchema<MediaInput> = object({
  id: string().uuid().nullable(),
  description: string().nullable(),
  copyright: string().nullable(),
  license: LicenseInputSchema.clone().nullable(),
  url: string().nonEmpty(),
  kind: mixed<MediaKind>().defined().oneOf(Object.values(MediaKind)),
  resolution: tuple([
    number().required().integer().positive(),
    number().required().integer().positive(),
  ]).notRequired(),
  weight: number().integer().nullable(),
})
  .strict(true)
  .noUnknown();
