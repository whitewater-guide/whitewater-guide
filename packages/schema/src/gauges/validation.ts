import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { mixed, object, string } from 'yup';

import type { GaugeInput } from '../__generated__/types';
import { PointInputSchema } from '../points';

export const GaugeInputSchema: ObjectSchema<GaugeInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  code: string().nonEmpty(),
  levelUnit: string().nullable(),
  flowUnit: string().nullable(),
  timezone: string().nullable(),
  location: PointInputSchema.clone().nullable(),
  requestParams: mixed().nullable(),
  url: string().url().nullable(),
  source: yupSchemas.refInput().required(),
})
  .strict(true)
  .noUnknown();
