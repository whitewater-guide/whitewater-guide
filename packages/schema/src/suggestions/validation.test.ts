import { createSafeValidator } from '@whitewater-guide/validation';

import type { SuggestionInput } from '../__generated__/types';
import {
  PhotoSuggestionInputSchema,
  SuggestionInputSchema,
} from './validation';

type TestValue = [string, SuggestionInput];
type IncorrectTestValue = [string, any];

describe('SuggestionInputSchema', () => {
  const validator = createSafeValidator(SuggestionInputSchema);

  const required: SuggestionInput = {
    section: { id: '45bcd1fc-b5de-11e9-a2a3-2a2ae2dbcce4' },
    description: 'descr',
  };

  const correct: SuggestionInput = {
    ...required,
    copyright: 'copyright',
    filename: null,
    resolution: null,
  };

  const correctValues: TestValue[] = [
    ['without media', correct],
    ['required fields only', required],
    [
      'with media',
      {
        ...correct,
        filename: 'fooo.jpg',
        resolution: [100, 100],
      },
    ],
  ];

  const incorrectValues: IncorrectTestValue[] = [
    ['all empty', { ...correct, description: null, copyright: null }],
    ['bad uuid', { ...correct, section: { id: 'fooo' } }],
    [
      'just media',
      {
        ...correct,
        description: null,
        copyright: null,
        filename: 'fooo.jpg',
        resolution: [100, 100],
      },
    ],
    ['no description and no media', { ...correct, description: null }],
    ['filename without resolution', { ...correct, filename: 'foo.jpg' }],
    ['resolution without filename', { ...correct, resolution: [100, 100] }],
    [
      'bad resolution',
      { ...correct, filename: 'foo.jpg', resolution: [100, 100, 100] },
    ],
    ['extra fields', { ...correct, foo: 'bar' }],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    expect(validator(value)).not.toBeNull();
    expect(validator(value)).toMatchSnapshot();
  });
});

describe('PhotoSuggestionInputSchema', () => {
  const validator = createSafeValidator(PhotoSuggestionInputSchema);

  const required: SuggestionInput = {
    section: { id: '45bcd1fc-b5de-11e9-a2a3-2a2ae2dbcce4' },
    filename: 'fooo.jpg',
    resolution: [100, 100],
  };

  const correct: SuggestionInput = {
    ...required,
    description: 'descr',
    copyright: 'copyright',
  };

  const correctValues: TestValue[] = [
    ['full', correct],
    ['required fields only', required],
    [
      'without description',
      {
        ...correct,
        description: null,
        copyright: null,
      },
    ],
  ];

  const incorrectValues: IncorrectTestValue[] = [
    [
      'all empty',
      {
        ...correct,
        filename: null,
        description: null,
        resolution: null,
        copyright: null,
      },
    ],
    ['no filename', { ...correct, filename: null }],
    ['no resolution', { ...correct, resolution: null }],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    expect(validator(value)).not.toBeNull();
    expect(validator(value)).toMatchSnapshot();
  });
});
