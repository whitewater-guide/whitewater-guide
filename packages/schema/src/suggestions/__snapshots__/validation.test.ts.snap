// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoSuggestionInputSchema should be invalid for all empty 1`] = `
{
  "filename": "yup:mixed.required",
  "resolution": "yup:mixed.required",
}
`;

exports[`PhotoSuggestionInputSchema should be invalid for no filename 1`] = `
{
  "filename": "yup:mixed.required",
}
`;

exports[`PhotoSuggestionInputSchema should be invalid for no resolution 1`] = `
{
  "resolution": "yup:mixed.required",
}
`;

exports[`SuggestionInputSchema should be invalid for all empty 1`] = `
{
  "description": "yup:mixed.required",
}
`;

exports[`SuggestionInputSchema should be invalid for bad resolution 1`] = `
{
  "resolution": {
    "key": "yup:array.max",
    "options": {
      "max": 2,
      "path": "resolution",
    },
  },
}
`;

exports[`SuggestionInputSchema should be invalid for bad uuid 1`] = `
{
  "section": {
    "id": {
      "key": "yup:string.uuid",
      "options": {
        "path": "section.id",
      },
    },
  },
}
`;

exports[`SuggestionInputSchema should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;

exports[`SuggestionInputSchema should be invalid for filename without resolution 1`] = `
{
  "resolution": "yup:mixed.required",
}
`;

exports[`SuggestionInputSchema should be invalid for just media 1`] = `
{
  "description": "yup:mixed.required",
}
`;

exports[`SuggestionInputSchema should be invalid for no description and no media 1`] = `
{
  "description": "yup:mixed.required",
}
`;

exports[`SuggestionInputSchema should be invalid for resolution without filename 1`] = `
{
  "filename": "yup:mixed.required",
}
`;
